ksmt-maxsat/src/main/kotlin/io/ksmt/solver/maxsat/KMaxSATSolver.kt

```diff
@@ -0,0 +1,328 @@
+package io.ksmt.solver.maxsat
+
+import io.ksmt.KContext
+import io.ksmt.expr.KExpr
+import io.ksmt.expr.KTrue
+import io.ksmt.solver.KModel
+import io.ksmt.solver.KSolver
+import io.ksmt.solver.KSolverConfiguration
+import io.ksmt.solver.KSolverStatus
+import io.ksmt.sort.KBoolSort
+import io.ksmt.utils.mkConst
+import kotlinx.coroutines.withTimeoutOrNull
+import kotlin.time.Duration
+
+class KMaxSATSolver<T>(private val ctx: KContext, private val solver: KSolver<T>) : KSolver<KSolverConfiguration>
+    where T : KSolverConfiguration {
+    private val scopeManager = MaxSATScopeManager()
+    private var softConstraints = mutableListOf<SoftConstraint>()
+
+    /**
+     * Assert softly an expression with weight (aka soft constraint) into solver.
+     *
+     * @see checkMaxSAT
+     * */
+    fun assertSoft(expr: KExpr<KBoolSort>, weight: Int) {
+        require(weight > 0) { "Soft constraint weight cannot be equal to $weight as it must be greater than 0" }
+
+        val softConstraint = SoftConstraint(expr, weight)
+        softConstraints.add(softConstraint)
+        scopeManager.incrementSoft()
+    }
+
+    /**
+     * Solve maximum satisfiability problem.
+     *
+     * @throws NotImplementedError
+     */
+    suspend fun checkMaxSAT(timeout: Duration = Duration.INFINITE): KMaxSATResult {
+        var hardConstraintsSatStatus = KSolverStatus.UNKNOWN
+        var currentMaxSATResult: Triple<KSolverStatus?, List<KExpr<KBoolSort>>, KModel?> =
+            Triple(null, listOf(), null)
+
+        solver.push()
+
+        val maxSATResult = withTimeoutOrNull(timeout) {
```

Comment:
This way of handling timeouts doesn't work, because `solver.check` is not a suspend function (has no suspension points).